packages/query-core/src/types.ts

```diff
@@ -46,7 +46,13 @@ export type QueryFunctionContext<
 
 export type InitialDataFunction<T> = () => T | undefined
 
-type NonFunctionGuard<T> = T extends Function ? never : T
+declare const queryCoreFunctionBrand: unique symbol
+type NotAFunction<T> = T & { [queryCoreFunctionBrand]?: never }
+declare global {
+  interface Function {
+    [queryCoreFunctionBrand]?: true
+  }
+}
```

Comment:
This is definitely somewhat unorthodox... and I'm battling myself if I should do it in XState. This is the best way I could find to prevent a type parameter from being inferred as a function when your property can be `T | (() => T)`. I don't have a source to back up my claim - but *similar* things were often~ suggested by TS team members in their issue tracker. I wish there would be a better way to do this but well, it is what it is (?) 🤷‍♂️

Reply:
do we even need the NonFunctionGuard? We also don't have this for `initialData`, which I think should be the same on type level

Reply:
It seems to behave slightly differently there - and I don't know why. I will investigate this but it might take some time before I get to it.